emcc.py

```diff
@@ -1632,9 +1643,12 @@ def do_minify(): # minifies the code. this is also when we do certain optimizati
       if opt_level >= 2:
         JSOptimizer.queue += ['asmLastOpts']
 
-      if shared.Settings.FINALIZE_ASM_JS: JSOptimizer.queue += ['last']
+      if shared.Settings.EMTERPRETIFY:
```

Comment:
Done! The reason for all of this was that when I tried running DFE right after registerizeHarder/registerize, the generated code didn't run. I took the easy route and put it right before the call to 'last', where it worked fine. But this was clashing with the EMTERPRETIFY mode since in EMTERPRETIFY mode, we perform the 'last' pass later and also defer do_minify. Hence the need for perform_last_opts. 

I dug in a bit, and it turns out that when whitespace minimization is enabled, uglify-js will not insert a ; after a function call if it happens to be the last function before an enclosing '}'. This was clashing with the asmLastOpts pass which modifies the AST and collapses a few scopes. As a result, we had a function call without a succeeding ';' followed by some more code, resulting in incorrect code.

It took a while to figure that out - but the simple trick after this was to disable whitespace minimization in the DFE pass (since the 'last' pass does this anyway). Once this was done, uglify was no longer silently removing ';' symbols, and everything worked fine. We do suffer a very slight loss in the perf of DFE with this change - I think performing DFE after function name minification is giving slightly better results, but I don't think its worth polluting the code like this.

So - emcc.py now looks much nicer. Fuzz testing is passing, so I think we should be good. I'll publish a new diff soon.